### ISSUE TITLE
Laravel and laravel-echo-server TokenMismatch

### ISSUE THREAD

**Issue by sidis405** [tlaverdure/laravel-echo-server#68]
Hi all.

Disclaimer. This works curectly with Pusher (so at least that's out of the way).

I started yesterday working on a project using Laravel echo and i wanted to use socket.io instead of pusher. So as suggested by Taylor on the official documentation, i went ahead and pulled in [https://github.com/tlaverdure/laravel-echo-server](https://github.com/tlaverdure/laravel-echo-server)

Doing this on Homestead.

Got everything working fine with public events.

The issues started with private channels.

As expected, and defined in the BroadcastServiceProvider, i need to send auth requests to /broadcasting/auth. The problem is i get a constant token mismatch exception.

Investigating further in Illuminate\Foundation\Http\Middleware\VerifyCsrfToken and specifically in the tokensMatch function, i notice that

```
$sessionToken = $request->session()->token();
```
and
```
 $token = $request->input('_token') ?: $request->header('X-CSRF-TOKEN');
```

have two different values. The one in the headers being the one of the authenticated session (so the correct one).

Moving further i was initially using redis as session store. Then i moved into file and noticed that for every request to /boadcasting/auth a new session was being generated.

In the end starting storing sessions into the database. Here's where the final clue came.

For every request coming from homestead (where node and laravel are, thus Echo is running) I am getting a new session generated with Homestead's ip 192.168.10.10 for every time i load the page. So naturally the sessions would not match but i can't figure out how to avoid it.

Additionally, even moving the application out of homestead yielded the same results.

-Basically echo server is unable to tell Laravel the currently logged in user.

I've cleared all caches, dumped autoload, cleared compiled, even rebuilt the vendor folder a couple of times.

At this point i have no idea anymore. I'd really appreciate if someone could shed some light.

Thanks in advance

**Comment by jonnywilliamson** [comment]
https://laravel.com/docs/5.3/broadcasting#introduction

> CSRF Token
> 
> Laravel Echo will need access to the current session's CSRF token. If available, Echo will pull the token from the Laravel.csrfToken JavaScript object. This object is defined in the  resources/views/layouts/app.blade.php layout that is created if you run the make:auth Artisan command. If you are not using this layout, you may define a meta tag in your application's head HTML element:
> 
`<meta name="csrf-token" content="{{ csrf_token() }}">`

Also in `bootstrap.js` assuming you're using Vue resource to send the requests should have

```
/**
 * We'll register a HTTP interceptor to attach the "CSRF" header to each of
 * the outgoing requests issued by this application. The CSRF middleware
 * included with Laravel will automatically verify the header's value.
 */

Vue.http.interceptors.push((request, next) => {
    request.headers['X-CSRF-TOKEN'] = Laravel.csrfToken;

    next();
});

```

**Comment by ctf0** [comment]
@jonnywilliamson the `Laravel.csrfToken` wont work unless u have the script part in the layout file to bind it so instead u can use
```js
Vue.http.interceptors.push((request, next) => {
    const csrfToken = document.querySelector('meta[name="csrf-token"]').getAttribute('content')
    request.headers.set('X-CSRF-TOKEN', csrfToken);
    next();
});
```

**Comment by jonnywilliamson** [comment]
@ctf0 Thanks! I wasn't having the issue it was @sidis405 .

Hopefully he got sorted.

**Comment by iddarohhi** [comment]
@sidis405 Did you find the solution? I'm having the same issue.

**Comment by barryvdh** [comment]
I just ignored the CSRF protection on the broadcasting route.. But CSRF protection works for me out of the box when I do this: ` <meta name="csrf-token" content="<?= csrf_token() ?>" />` in the `<head>`, only problem is that it expires after inactivity

**Comment by barryvdh** [comment]
Is you authentication domain exactly the same as the app you're viewing? Can you verify that the Cookies are sent with your Socket (eg. open the WebSocket tab in Chrome dev tools and view the request headers)

**Comment by ismailkriam** [comment]
i have the same problem.
i think the laravel-echo-server sent the request with new session.
and if i disable  CSRF verification middleware  ,  the request->user() is return null.
this is the problem.

**Comment by barryvdh** [comment]
See my questions in previous comment. 
If no valid cookies are sent, the session id isn't correctly passed through and will generate a new session.

**Comment by ismailkriam** [comment]
please see the snapshot. 
![snapshot](https://cloud.githubusercontent.com/assets/8618561/22167545/c5c0cd38-df6f-11e6-87d7-36a49356a1e0.PNG)

**Comment by barryvdh** [comment]
So no laravel session cookies. What if you visit the website also on http://localhost instead of 127.0.0.1?

**Comment by ismailkriam** [comment]
thanks for your help @barryvdh , the problem fixed .
i'm must access the applecation used localhost not the ip .
this is the problem.

**Comment by barryvdh** [comment]
Yes you must connect to the web socket over the same domain as the app to share cookies, or use state less login with jwt headers or something.

**Comment by barryvdh** [comment]
Tip, try something like this in your client side code: 


window.Echo = new Echo({
    broadcaster: 'socket.io',
    host: window.location.hostname + ':6001'
});

**Comment by mdixon18** [comment]
I am getting a TokenMismatch too, I have gone into Echo.js to see what happens when the request is sent and i logged out the headers. It is adding in the CSRF token itself which is great. 

However, when the auth tries to authenticate I am getting 

> Client can not be authenticated, got HTTP status 419

Which looking at HttpException refers to TokenMismatch. Is there any reason why even with the CSRF token in place that i would get a mismatch?

```
Authorization:"Bearer blahblah"
X-CSRF-TOKEN:"ix7T4lZA9mfCxrd9jDAbeGvSglMeoge7X7fG8LLi"
```

**Comment by tlaverdure** [comment]
Hey there, this issue was opened a while ago. I'm going to close this issue, if this issue still exists, please open a new issue or open a Pull Request.

**Comment by marothyzsolt** [comment]
I have same problem. TokenMismatchException, and if i disable csrf the request user is null. I'm using different domain on websocket, and the webserver, I'm using a subdomain with different IP address for the websocket  (https://ws.example.com:443). I must use another domain on port 443, because some on some networks has been disabled every port except the main ports (HTTP, HTTPS, IMAP, POP3, etc).

Update: I solved the problem: changed session.php 'domain' changed to: .example.com

**Comment by sidis405** [comment]
It's funny when years later you have the same problem and you bump on your own issue only to find a solution from 3 hours ago. Thank you @marothyzsolt 💯

**Comment by efriandika** [comment]
thanks @marothyzsolt .. You save my live in 2020.. :D

**Comment by david95thinkcode** [comment]
I've faced this issue on my local dev environment. 
In my case I solved it by replacing **127.0.0.1** by **localhost**.
<img width="1552" alt="Capture d’écran 2020-02-03 à 16 03 17" src="https://user-images.githubusercontent.com/24503308/73664259-6535f200-469f-11ea-8312-fb5c7eebe95e.png">


Thanks to @barryvdh and @marothyzsolt , your comments help me a so much.

**Comment by SepehrMoafi** [comment]
i set localhost and ip and  APP_URL  and host to 0.0.0.0 thhen i send CSRF token 
some time may thay have diffrent destination that why i set them in one (0.0.0.0) and it works !
and csrf like this 
```
    window.Echo = new MyEcho({
        broadcaster: 'socket.io',
        host: '0.0.0.0:6001', // this is laravel-echo-server host
        auth:{
            headers: {
            'X-CSRF-TOKEN': '{{csrf_token()}}'
        }}

    });
```
